airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py

```diff
@@ -97,59 +95,51 @@ def __post_init__(self, parameters: Mapping[str, Any]):
     def get_stream_state(self) -> StreamState:
         return {self.cursor_field.eval(self.config): self._cursor} if self._cursor else {}
 
-    def update_cursor(self, stream_slice: StreamSlice, last_record: Optional[Record] = None):
+    def set_initial_state(self, stream_state: StreamState) -> None:
+        """
+        Cursors are not initialized with their state. As state is needed in order to function properly, this method should be called
+        before calling anything else
+
+        :param stream_state: The state of the stream as returned by get_stream_state
+        """
+        self._cursor = stream_state.get(self.cursor_field.eval(self.config)) if stream_state else None
+
+    def update_state(self, stream_slice: StreamSlice, last_record: Record) -> None:
         """
         Update the cursor value to the max datetime between the last record, the start of the stream_slice, and the current cursor value.
-        Update the cursor_end value with the stream_slice's end time.
 
         :param stream_slice: current stream slice
         :param last_record: last record read
         :return: None
         """
-        stream_slice_value = stream_slice.get(self.cursor_field.eval(self.config))
-        stream_slice_value_end = stream_slice.get(self.partition_field_end.eval(self.config))
         last_record_value = last_record.get(self.cursor_field.eval(self.config)) if last_record else None
-        cursor = None
-        if stream_slice_value and last_record_value:
-            cursor = max(stream_slice_value, last_record_value)
-        elif stream_slice_value:
-            cursor = stream_slice_value
-        else:
-            cursor = last_record_value
-        if self._cursor and cursor:
-            self._cursor = max(cursor, self._cursor)
-        elif cursor:
-            self._cursor = cursor
-        if self.partition_field_end:
-            self._cursor_end = stream_slice_value_end
-
-    def stream_slices(self, sync_mode: SyncMode, stream_state: Mapping[str, Any]) -> Iterable[Mapping[str, Any]]:
+
+        possible_cursor_values = list(filter(lambda item: item, [last_record_value, self._cursor]))
+        self._cursor = max(possible_cursor_values) if possible_cursor_values else None
```

Comment:
Note that there is a bug here if the datetime format can't be ordered by alphabetical order since we compare strings. For example, datetime format `%d-%m-%Y` will not work as expected. I have created an issue for this https://github.com/airbytehq/airbyte/issues/27447